src/lib/Microsoft.Fx.Portability.Azure/Storage/StorageBlob.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Fx.Portability.ObjectModel;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Fx.Portability.Azure.Storage
+{
+    public class StorageBlob
+    {
+        private static readonly Version StorageVersion = new Version(1, 0, 0, 0);
+
+        private readonly CloudStorageAccount _storageAccount;
+
+        public StorageBlob(CloudStorageAccount storageAccount)
+        {
+            _storageAccount = storageAccount;
+        }
+
+        /// <summary>
+        /// Saves the given analysis request to the current month's container
+        /// keyed by its uniqueId as shown below:
+        /// * container{year}{month}/{day}/{uniqueId}
+        /// * Example:
+        ///     container201705/10/1234-ABCD represents the request 1234-ABCD
+        ///     that was submitted on May 10, 2017.
+        /// </summary>
+        /// <param name="uniqueId">Guid for the analysis request</param>
+        /// <param name="request">The contents of the request.</param>
+        /// <returns></returns>
+        public async Task<bool> SaveToBlobAsync(string uniqueId, AnalyzeRequest request)
+        {
+            DateTime currentDate = DateTime.Now;
+            string containerName = GetCurrentMonthContainerName(currentDate);
+            CloudBlobContainer container = await GetBlobContainerAsync(containerName);
+
+            if (container == null)
+            {
+                return false;
+            }
+
+            byte[] content = request.SerializeAndCompress();
+            var containerPath = $"{currentDate.Day.ToString("00", CultureInfo.InvariantCulture)}/{uniqueId}";
+
+            CloudBlockBlob blob = container.GetBlockBlobReference(containerPath);
+            blob.Metadata.Add("Version", request.Version.ToString(CultureInfo.InvariantCulture));
+            await blob.UploadFromByteArrayAsync(content, 0, content.Length);
+            await blob.SetMetadataAsync();
+            return true;
+        }
+
+        public async Task<AnalyzeRequest> RetrieveFromBlobAsync(string uniqueId, string containerName)
+        {
+            CloudBlobContainer container = await GetBlobContainerAsync(containerName);
+            CloudBlockBlob blob = container.GetBlockBlobReference(uniqueId);
+
+            using (var blobStream = new MemoryStream())
+            {
+                await blob.DownloadToStreamAsync(blobStream);
+
+                var blobData = blobStream.ToArray();
```

Comment:
IIRC, there is be a JsonDeserializer we have in Microsoft.Fx.Portability. available that takes a stream instead of an array. It would be good not to actualize this value and deserialize the stream instead.

Reply:
Good suggestion. I didn't know that.  Fixed.

---
In reply to: [178098527](https://github.com/Microsoft/dotnet-apiport/pull/598#discussion_r178098527) [](ancestors = 178098527)